### PR TITLE
(PUP-6344) Capture the time of the entire run as the total

### PR DESCRIPTION
**Pull request by kris-bosland**
Without this change, the reported total time is incorrect, since it was double counting.
With this change, the total is now reported from measuring the time of the run operation.

### CODE REVIEW

File: lib/puppet/transaction/report.rb [hunk 1]
```diff
@@ -351,9 +351,9 @@ def summary
 
       report[key].keys.sort { |a,b|
         # sort by label
-        if a == TOTAL
+        if a == :total
```
Comment (melissa) [inline lib/puppet/transaction/report.rb]: what happened here? with this change, we no longer have `total` appearing at the end of the report

```
Time:
         Schedule: 0.00
          Package: 0.00
   Property list key: 0.00
          Printer: 0.01
             Cron: 0.01
             File: 0.07
          Service: 0.10
   Profile manager: 0.31
   Convert catalog: 0.46
             Exec: 1.66
   Transaction evaluation: 11.75
   Catalog application: 11.86
         Last run: 1522792065
      Plugin sync: 3.64
            Total: 30.50
   Config retrieval: 4.43
   Node retrieval: 4.54
   Fact generation: 5.55
        Pe anchor: 0.00
       Filebucket: 0.00
``

is there a reason you reverted this change?

File: spec/unit/transaction/report_spec.rb [hunk 2]
```diff
@@ -392,15 +376,6 @@ def add_statuses(count, type = :file)
...
       end
-
-      it "should have a total time" do
-        add_statuses(3, :file) do |status|
-          status.evaluation_time = 1.25
-        end
-        @report.add_times :config_retrieval, 0.5
-        @report.finalize_report
-        expect(metric(:time, "total")).to eq(4.25)
-      end
```
Comment (melissa) [inline spec/unit/transaction/report_spec.rb]: Why did you remove these tests? It seems like we do still want to have a total time in the report, even though it's no longer being calculated in the same way. Maybe we just want to check that `total` is present in the report?

Comment (kris-bosland) [reply spec/unit/transaction/report_spec.rb]: I wasn't able to get this test to react to adding/removing the new total time entry here.

File: spec/unit/configurer_spec.rb [hunk 3]
```diff
@@ -359,6 +359,13 @@
       @agent.apply_catalog(@catalog, {:report => report})
     end
 
+    it "should record a total run time" do
+      report = Puppet::Transaction::Report.new
+      @catalog.stubs(:apply).with(:report => report)
+      report.expects(:add_times).with(:total, anything())
```
Comment (kris-bosland) [inline spec/unit/configurer_spec.rb]: I was able to get this test to react to adding/removing the total time entry here.

Comment (kris-bosland) [inline spec/unit/configurer_spec.rb]: I was able to get this test to react to adding/removing the new total time entry here.

**Comment by kris-bosland** [comment]
Hi @melissa, some testing has moved, but I think this gives the coverage and results that you were looking for.  Thanks!

File: lib/puppet/configurer.rb [hunk 4]
```diff
@@ -247,132 +243,144 @@ def run(options = {})
...
+          return nil
+        end
+      end #thinmark
+    ensure
+      report.add_times(:total, run_internal_time)
+      report.finalize_report
 
-      execute_prerun_command or return nil
+      exit_status = report.exit_status if prerun_command_successful
+      exit_status = nil unless execute_postrun_command
```
Comment (melissa) [inline lib/puppet/configurer.rb]: I'm having a little trouble following this logic. I know it's handling the logic pre

I *think* it might be a little easier to follow if it were written more like

```
if prerun_command_successful
  exit_status = report.exit status
  execute_postrun_command or return nil
  exit_status
end
```

but I'm not really sure the best way to change this. I just know when we look back at this later, it's going to take us a minute to parse out the what and why of this

File: spec/unit/transaction/report_spec.rb (hunk 2 above)
Comment (melissa) [reply spec/unit/transaction/report_spec.rb]: Yeah, we probably don't have a total time in the report in these. This is fine

File: spec/unit/transaction/report_spec.rb [hunk 5]
```diff
@@ -488,6 +463,7 @@ def add_statuses(count, type = :file)
       trans = catalog.apply
 
       @report = trans.report
+      @report.add_times(:total, "8675") #Report total is now measured, not calculated.
```
Comment (melissa) [inline spec/unit/transaction/report_spec.rb]: Why do we need to add total here if we're not checking it elsewhere in this test? Was this just leftover from earlier testing?

File: lib/puppet/configurer.rb (hunk 4 above)
Comment (melissa) [reply lib/puppet/configurer.rb]: This is definitely a nitpick, and shouldn't block merging this in if we're worried about time

File: spec/unit/transaction/report_spec.rb (hunk 5 above)
Comment (kris-bosland) [reply spec/unit/transaction/report_spec.rb]: It is actually needed in the [later test](https://github.com/kris-bosland/puppet/blob/5c092467789093d87228c1f471a469d314e3ae4f/spec/unit/transaction/report_spec.rb#L513), we have to add a value directly because the total is no longer adding the values that are already added in the test.

File: lib/puppet/configurer.rb (hunk 4 above)
Comment (jhelwig) [reply lib/puppet/configurer.rb]: @melissa So, the problem with that `if` form is that we always want `execute_postrun_command` to run, even if the prerun command failed, so we can't have execute_postrun_command in the body of that if. I'm not sure of a cleaner way of setting up the exit status where we have both the prerun, and postrun commands run regardless of other success/failure.
